In [ ]:
import json
import os, re
from collections import Counter

# Data loading code

Counter(labels)


Counter({1: 195, 0: 168})

To replicate this study you need to obtain ada embeddings from OpenAI using their API and load  as them as embeddings variable (list of embeddings) used in the code below:

Counter({0: 168, 1: 195})



In [ ]:
from sklearn.model_selection import KFold
from sklearn.metrics import classification_report
import numpy as np
from sklearn.svm import NuSVC
from collections import Counter
import pickle

#for kernel_type in ['rbf', 'linear']:
kernel_type = 'rbf'

labels = np.array(labels)
embeddings = np.array(embeddings)

true_labels = [] # ground truth
predictions = [] # model predictions

runtime_name = 'ada'

kf = KFold(n_splits=10, shuffle=True, random_state=42)
for train_index, test_index in kf.split(labels):

  X_train, X_test = embeddings[train_index], embeddings[test_index]
  y_train, y_test = labels[train_index], labels[test_index]

  print(Counter(y_train))

  clf = NuSVC(nu=0.5, kernel=kernel_type)

  model = clf.fit(X_train, y_train.ravel())
  y_pred = model.predict(X_test).tolist()

  predictions.append( y_pred )
  true_labels.append( y_test )


print( classification_report(np.concatenate(true_labels).ravel().tolist(), np.concatenate(predictions).ravel().tolist() ) )


Counter({1: 177, 0: 149})
Counter({1: 172, 0: 154})
Counter({1: 178, 0: 148})
Counter({1: 179, 0: 148})
Counter({1: 175, 0: 152})
Counter({1: 176, 0: 151})
Counter({1: 177, 0: 150})
Counter({1: 173, 0: 154})
Counter({1: 175, 0: 152})
Counter({1: 173, 0: 154})
              precision    recall  f1-score   support

           0       0.80      0.88      0.84       168
           1       0.88      0.81      0.84       195

    accuracy                           0.84       363
   macro avg       0.84      0.84      0.84       363
weighted avg       0.84      0.84      0.84       363



Print the number of parameters of the SVM classifier:

In [ ]:
def n_params(model):
    """Return total number of parameters in a
    Scikit-Learn model.

    This works for the following model types:
     - sklearn.neural_network.MLPClassifier
     - sklearn.neural_network.MLPRegressor
     - sklearn.linear_model.LinearRegression
     - and maybe some others
    """
    return (
            sum([a.size for a in model.dual_coef_]) +
            model.support_vectors_.shape[0] * model.support_vectors_.shape[1]
            )

n_params(clf)

324307